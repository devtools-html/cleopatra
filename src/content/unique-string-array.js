// @flow
import type { IndexIntoStringTable } from '../common/types/profile';

export class UniqueStringArray {

  _array: string[]
  _stringToIndex: Map<string, IndexIntoStringTable>

  constructor(originalArray: string[] = []) {
    this._array = originalArray.slice(0);
    this._stringToIndex = new Map(originalArray.map((s, i) => [s, i]));
  }

  getString(index: IndexIntoStringTable): string {
    if (!(index in this._array)) {
      throw new Error(`index ${index} not in UniqueStringArray`);
    }
    return this._array[index];
  }

  indexForString(s: string): IndexIntoStringTable {
    let index = this._stringToIndex.get(s);
    if (index === undefined) {
      index = this._array.length;
      this._stringToIndex.set(s, index);
      this._array.push(s);
    }
    return index;
  }

  serializeToArray(): string[] {
    return this._array.slice(0);
  }
}
