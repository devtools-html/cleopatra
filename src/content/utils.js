// @flow
type ValidWeakMapKey = Object | [];

/**
 * Memoize a function's results using a WeakMap and its first parameter.
 */
export function weakMemoize(fn: (ValidWeakMapKey, ...any) => any) {
  const memo = new WeakMap();
  return function (key: ValidWeakMapKey) {
    const cachedValue = memo.get(key);
    if (cachedValue === undefined) {
      const value = fn.apply(null, arguments);
      memo.set(key, value);
      return value;
    }
    return cachedValue;
  };
}
