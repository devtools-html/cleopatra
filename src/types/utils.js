/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

/*
 * A not-void type
 * https://flow.org/try/#0C4TwDgpgBAcg9sAanAlgEygXigKClAHygDsBXAWwCMIAnPQqAZ2BpWIHN6jK44AbCAENiXEqT59RAQRo1BIADzCQAPlEB5SgCsIAY2ABuHDgBmpYvpRxiUMxYUAVAFywEydCoAUAN0F9SEC4OAJRBUADe9DQQwKQ0Nr7+EEYAvsZ2up4AjMFGGZ4ARMysHAW5puaZ4ba8LgDklII0dVAp5flkEu2VngDaWQA0UABMALrdFp7maBAmbBBoE5nlQA
 */
export type NotVoidOrNull = number | string | boolean | Array<any> | Object;

export type ExtractReturnType = <V>((...args: any[]) => V) => V;

/**
 * This type serves as documentation for how an array is meant to be used, but does
 * not support type checking. We often use an Array instead of a Map to translate
 * one type of index into another type of index. This is similar to how we use the
 * Map<K,V> type, but with the Array.
 */
// eslint-disable-next-line no-unused-vars
export type IndexedArray<_IndexType, Value> = Array<Value>;

/**
 * This is a utility type that extracts the return type of a function.
 */
export type $ReturnType<Fn> = $Call<ExtractReturnType, Fn>;
