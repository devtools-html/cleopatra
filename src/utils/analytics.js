/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

/**
 * Document Google Analytics API that is used in the project. These definitions
 * can be updated as more features are used.
 *
 * https://developers.google.com/analytics/devguides/collection/analyticsjs/pages
 */
type GAEvent = {|
  hitType: 'event',
  // Specifies the event category. Must not be empty
  eventCategory: string,
  eventAction: string,
  eventLabel?: string,
  eventValue?: number,
|};

type GAPageView = {|
  hitType: 'pageview',
  page: string,
|};

type GATiming = {|
  hitType: 'timing',
  timingCategory: string,
  timingVar: string,
  timingValue: number,
  timingLabel?: string,
|};

type GAPayload = GAEvent | GAPageView | GATiming;

export type GoogleAnalytics = ('send', GAPayload) => {};

export function sendAnalytics(payload: GAPayload) {
  const ga: ?GoogleAnalytics = self.ga;
  if (ga) {
    ga('send', payload);
  }
}
