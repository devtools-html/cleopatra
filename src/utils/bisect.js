// This file originally comes from https://github.com/3rd-Eden/node-bisection/, but has been
// imported and changed to fix some bugs.

/*
    Copyright (c) 2010/2011 Arnout Kazemier,3rd-Eden
    Permission is hereby granted, free of charge, to any person obtaining a copy
    of this software and associated documentation files (the "Software"), to deal
    in the Software without restriction, including without limitation the rights
    to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
    copies of the Software, and to permit persons to whom the Software is
    furnished to do so, subject to the following conditions:
    The above copyright notice and this permission notice shall be included in
    all copies or substantial portions of the Software.
    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
    THE SOFTWARE. 
*/

/*
    The two functions here, bisectionLeft and bisectionRight, return the index where a new element
    would be inserted, respectively at the left or at the right of elements with the same value. 
*/

// @flow

export function bisectionRight(
  array: number[] | $TypedArray,
  x: number,
  low?: number,
  high?: number
): number {
  low = low || 0;
  high = high || array.length;

  if (low < 0 || low > array.length || high < 0 || high > array.length) {
    throw new TypeError("low and high must lie within the array's range");
  }

  while (low < high) {
    const mid = (low + high) >> 1;

    if (x < array[mid]) {
      high = mid;
    } else {
      low = mid + 1;
    }
  }

  return low;
}

export function bisectionLeft(
  array: number[] | $TypedArray,
  x: number,
  low?: number,
  high?: number
): number {
  low = low || 0;
  high = high || array.length;

  if (low < 0 || low > array.length || high < 0 || high > array.length) {
    throw new TypeError("low and high must lie within the array's range");
  }

  while (low < high) {
    const mid = (low + high) >> 1;

    if (x <= array[mid]) {
      high = mid;
    } else {
      low = mid + 1;
    }
  }

  return low;
}

export function bisectionLeftBy<T>(
  array: T[],
  f: (T) => number, // < 0 if arg is before needle, > 0 if after, === 0 if same
  low?: number,
  high?: number
): number {
  low = low || 0;
  high = high || array.length;

  if (low < 0 || low > array.length || high < 0 || high > array.length) {
    throw new TypeError("low and high must lie within the array's range");
  }

  while (low < high) {
    const mid = (low + high) >> 1;

    if (f(array[mid]) >= 0) {
      high = mid;
    } else {
      low = mid + 1;
    }
  }

  return low;
}

export function bisectionLeftByKey<T>(
  array: T[],
  x: number,
  toKey: (T) => number,
  low?: number,
  high?: number
): number {
  low = low || 0;
  high = high || array.length;

  if (low < 0 || low > array.length || high < 0 || high > array.length) {
    throw new TypeError("low and high must lie within the array's range");
  }

  while (low < high) {
    const mid = (low + high) >> 1;

    if (x <= toKey(array[mid])) {
      high = mid;
    } else {
      low = mid + 1;
    }
  }

  return low;
}

// This is the same as bisectionLeftByKey but uses string as the key type.
// If you can find a way to make Flow accept a single function that handles
// both string and number keys, please remove this duplication.
export function bisectionLeftByStrKey<T>(
  array: T[],
  x: string,
  toKey: (T) => string,
  low?: number,
  high?: number
): number {
  low = low || 0;
  high = high || array.length;

  if (low < 0 || low > array.length || high < 0 || high > array.length) {
    throw new TypeError("low and high must lie within the array's range");
  }

  while (low < high) {
    const mid = (low + high) >> 1;

    if (x <= toKey(array[mid])) {
      high = mid;
    } else {
      low = mid + 1;
    }
  }

  return low;
}

/*
 * bisectLowerBound(), bisectUpperBound(), and bisectEqualRange() are equivalent to
 * std::lower_bound(), std::upper_bound(), and std::equal_range() respectively.
 *
 * bisectLowerBound() returns an index pointing to the first element in the range
 * in which each element is considered *not less than* the given value passed
 * via |aCompare|, or the length of |aContainer| if no such element is found.
 *
 * bisectUpperBound() returns an index pointing to the first element in the range
 * in which each element is considered *greater than* the given value passed
 * via |aCompare|, or the length of |aContainer| if no such element is found.
 *
 * bisectEqualRange() returns a range [first, second) containing all elements are
 * considered equivalent to the given value via |aCompare|.  If you need
 * either the first or last index of the range, bisectLowerBound() or bisectUpperBound(),
 * which is slightly faster than bisectEqualRange(), should suffice.
 *
 * Example (another example is given in TestBinarySearch.cpp):
 *
 *   Vector<const char*> sortedStrings = ...
 *
 *   struct Comparator {
 *     const nsACString& mStr;
 *     explicit Comparator(const nsACString& aStr) : mStr(aStr) {}
 *     int32_t operator()(const char* aVal) const {
 *       return Compare(mStr, nsDependentCString(aVal));
 *     }
 *   };
 *
 *   auto bounds = bisectEqualRange(sortedStrings, 0, sortedStrings.length(),
 *                            Comparator("needle I'm looking for"_ns));
 *   printf("Found the range [%zd %zd)\n", bounds.first(), bounds.second());
 *
 */
export function bisectLowerBound(
  array: number[] | $TypedArray,
  f: (number) => number, // < 0 if arg is before needle, > 0 if after, === 0 if same
  low?: number,
  high?: number
): number {
  low = low || 0;
  high = high || array.length;

  if (low < 0 || low > array.length || high < 0 || high > array.length) {
    throw new TypeError("low and high must lie within the array's range");
  }

  while (high !== low) {
    const middle = (low + high) >> 1;
    const result = f(array[middle]);

    // The range returning from bisectLowerBound does include elements
    // equivalent to the given value i.e. f(element) == 0
    if (result >= 0) {
      high = middle;
    } else {
      low = middle + 1;
    }
  }

  return low;
}

export function bisectUpperBound(
  array: number[] | $TypedArray,
  f: (number) => number, // < 0 if arg is before needle, > 0 if after, === 0 if same
  low?: number,
  high?: number
): number {
  low = low || 0;
  high = high || array.length;

  if (low < 0 || low > array.length || high < 0 || high > array.length) {
    throw new TypeError("low and high must lie within the array's range");
  }

  while (high !== low) {
    const middle = (low + high) >> 1;
    const result = f(array[middle]);

    // The range returning from bisectUpperBound does NOT include elements
    // equivalent to the given value i.e. f(element) == 0
    if (result > 0) {
      high = middle;
    } else {
      low = middle + 1;
    }
  }

  return high;
}

export function bisectEqualRange(
  array: number[] | $TypedArray,
  f: (number) => number, // < 0 if arg is before needle, > 0 if after, === 0 if same
  low?: number,
  high?: number
): [number, number] {
  low = low || 0;
  high = high || array.length;

  if (low < 0 || low > array.length || high < 0 || high > array.length) {
    throw new TypeError("low and high must lie within the array's range");
  }

  while (high !== low) {
    const middle = (low + high) >> 1;
    const result = f(array[middle]);

    if (result > 0) {
      high = middle;
    } else if (result < 0) {
      low = middle + 1;
    } else {
      return [
        bisectLowerBound(array, f, low, middle),
        bisectUpperBound(array, f, middle + 1, high),
      ];
    }
  }

  return [low, high];
}
