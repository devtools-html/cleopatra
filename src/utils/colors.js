/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

/**
 * These are the colors from Photon. They are inlined to provide easy access. If updating
 * please change the CSS variables as well.
 *
 * Firefox Colors v1.0.3
 * https://github.com/FirefoxUX/design-tokens/tree/master/colors
 */

export const MAGENTA_50 = '#ff1ad9';
export const MAGENTA_60 = '#ed00b5';
export const MAGENTA_70 = '#b5007f';
export const MAGENTA_80 = '#7d004f';
export const MAGENTA_90 = '#440027';

export const PURPLE_50 = '#9400ff';
export const PURPLE_60 = '#8000d7';
export const PURPLE_70 = '#6200a4';
export const PURPLE_80 = '#440071';
export const PURPLE_90 = '#25003e';

export const BLUE_40 = '#45a1ff';
export const BLUE_50 = '#0a84ff';
export const BLUE_60 = '#0060df';
export const BLUE_70 = '#003eaa';
export const BLUE_80 = '#002275';
export const BLUE_90 = '#000f40';

export const TEAL_50 = '#00feff';
export const TEAL_60 = '#00c8d7';
export const TEAL_70 = '#008ea4';
export const TEAL_80 = '#005a71';
export const TEAL_90 = '#002d3e';

export const GREEN_50 = '#30e60b';
export const GREEN_60 = '#12bc00';
export const GREEN_70 = '#058b00';
export const GREEN_80 = '#006504';
export const GREEN_90 = '#003706';

export const YELLOW_50 = '#ffe900';
export const YELLOW_60 = '#d7b600';
export const YELLOW_70 = '#a47f00';
export const YELLOW_80 = '#715100';
export const YELLOW_90 = '#3e2800';

export const RED_50 = '#ff0039';
export const RED_60 = '#d70022';
export const RED_70 = '#a4000f';
export const RED_80 = '#5a0002';
export const RED_90 = '#3e0200';

export const ORANGE_50 = '#ff9400';
export const ORANGE_60 = '#d76e00';
export const ORANGE_70 = '#a44900';
export const ORANGE_80 = '#712b00';
export const ORANGE_90 = '#3e1300';

export const GREY_10 = '#f9f9fa';
export const GREY_20 = '#ededf0';
export const GREY_30 = '#d7d7db';
export const GREY_40 = '#b1b1b3';
export const GREY_50 = '#737373';
export const GREY_60 = '#4a4a4f';
export const GREY_70 = '#38383d';
export const GREY_80 = '#2a2a2e';
export const GREY_90 = '#0c0c0d';

export const INK_70 = '#363959';
export const INK_80 = '#202340';
export const INK_90 = '#0f1126';
