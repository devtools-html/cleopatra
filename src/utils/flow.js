/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

/**
 * This file contains utils that help Flow understand things better. Occasionally
 * statements can be logically equivalent, but Flow infers them in a specific way. Most
 * of the time tweaks can be done by editing the type system, but occasionally functions
 * are needed to get the desired result.
 */

/**
 * This function can be run as the default arm of a switch statement to ensure exhaustive
 * checking of a given type. It relies on an assumption that a given case will not be
 * the string 'Error: non exhaustive switch found.' This assumption generates more
 * readable errors. Flow will only generate an error if it's possible to get there
 * within the type system.
 */
export function unexpectedCase(notValid: empty): void {
  throw new Error(`Unexpected case ${notValid}`);
}

/**
 * Immutably update an object through Object.assign, but retain the original
 * type information of the object. Flow will occasionally throw errors when
 * inferring what is going on with Object.assign.
 */
export function immutableUpdate<T: Object>(object: T, ...rest: Object[]): T {
  return Object.assign({}, object, ...rest);
}
