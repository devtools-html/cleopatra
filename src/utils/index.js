/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

/**
 * Firefox has issues switching quickly between fill style colors, as the CSS color
 * is fully parsed each time it is set. As a mitigation, provide a class that only
 * switches the color when it's really needed.
 */
export class FastFillStyle {
  _ctx: CanvasRenderingContext2D;
  _previousFillColor: string;

  constructor(ctx: CanvasRenderingContext2D) {
    this._ctx = ctx;
    this._previousFillColor = '';
  }

  set(fillStyle: string) {
    if (fillStyle !== this._previousFillColor) {
      this._ctx.fillStyle = fillStyle;
      this._previousFillColor = fillStyle;
    }
  }
}

/**
 * Perform a simple shallow object equality check.
 */
export function objectShallowEquals(a: Object, b: Object): boolean {
  let aLength = 0;
  let bLength = 0;
  for (const key in a) {
    if (a.hasOwnProperty(key)) {
      aLength++;
      if (a[key] !== b[key]) {
        return false;
      }
    }
  }
  for (const key in b) {
    if (b.hasOwnProperty(key)) {
      bLength++;
    }
  }
  return aLength === bLength;
}
