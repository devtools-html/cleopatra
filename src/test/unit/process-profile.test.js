/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

import {
  extractFuncsAndResourcesFromFrameLocations,
  processProfile,
  serializeProfile,
  unserializeProfileOfArbitraryFormat,
  sanitizePII,
} from '../../profile-logic/process-profile';
import { UniqueStringArray } from '../../utils/unique-string-array';
import {
  createGeckoProfile,
  createGeckoCounter,
} from '../fixtures/profiles/gecko-profile';
import { ensureExists } from '../../utils/flow';
import type { RemoveProfileInformation } from '../../types/profile-derived';

describe('extract functions and resource from location strings', function() {
  // These location strings are turned into the proper funcs.
  const locations = [
    // Extract unsymbolicated memory and match them to libraries.
    '0xc0ff33',
    '0xe0ff33',

    // Extract C++ function names and resources
    'cppFunction1 (in c++ resource name1) + 123',
    'cppFunction2 (in c++ resource name2) (234:345)',
    'cppFunction3 (in c++ resource name2)',

    // Extract JS functions URL information
    'jsFunction1 (http://script.com/one.js:456:1)',
    'http://script.com/one.js:456:1',

    // Extension locations
    'moz-extension://bf3bb73c-919c-4fef-95c4-070a19fdaf85/background.js:1:0',
    'moz-extension://fa2edf9c-c45f-4445-b819-c09e3f2d58d5/content.js:1:0',
    'backgroundFunction (moz-extension://bf3bb73c-919c-4fef-95c4-070a19fdaf85/background.js:2:1)',
    'contentfunction (moz-extension://fa2edf9c-c45f-4445-b819-c09e3f2d58d5/content.js:2:1)',

    // Something unknown
    'mysterious location',
  ];
  const libs = [
    // This library will match the '0xc0ff33' location.
    {
      start: 0xc00000,
      end: 0xd00000,
      offset: 0,
      arch: '',
      name: 'No symbols library',
      path: '',
      debugName: '',
      debugPath: '',
      breakpadId: '',
    },
    // This library will match the '0xe0ff33' location, and it has an offset.
    {
      start: 0xe01000,
      end: 0xf00000,
      offset: 0x1000,
      arch: '',
      name: 'No symbols library',
      path: '',
      debugName: '',
      debugPath: '',
      breakpadId: '',
    },
  ];
  const stringTable = new UniqueStringArray();
  const locationIndexes = locations.map(location =>
    stringTable.indexForString(location)
  );
  const extensions = {
    baseURL: [
      'moz-extension://bf3bb73c-919c-4fef-95c4-070a19fdaf85/',
      'moz-extension://fa2edf9c-c45f-4445-b819-c09e3f2d58d5/',
    ],
    id: ['geckoprofiler@mozilla.com', 'screenshots@mozilla.org'],
    name: ['Gecko Profiler', 'Firefox Screenshots'],
    length: 2,
  };

  it('extracts the information for all different types of locations', function() {
    const [
      funcTable,
      resourceTable,
      frameFuncs,
    ] = extractFuncsAndResourcesFromFrameLocations(
      locationIndexes,
      locationIndexes.map(() => false),
      stringTable,
      libs,
      extensions
    );

    expect(
      frameFuncs.map((funcIndex, locationIndex) => {
        // Map all the results into a human readable object for easy snapshotting.
        const locationName = locations[locationIndex];

        const funcName = stringTable.getString(funcTable.name[funcIndex]);
        const resourceIndex = funcTable.resource[funcIndex];
        const address = funcTable.address[funcIndex];
        const isJS = funcTable.isJS[funcIndex];
        const fileNameIndex = funcTable.fileName[funcIndex];
        const fileName =
          fileNameIndex === null ? null : stringTable.getString(fileNameIndex);
        const lineNumber = funcTable.lineNumber[funcIndex];
        const columnNumber = funcTable.columnNumber[funcIndex];

        let libIndex, resourceName, host, resourceType;
        if (resourceIndex === -1) {
          resourceName = null;
          host = null;
          resourceType = null;
        } else {
          const hostStringIndex = resourceTable.host[resourceIndex];
          libIndex = resourceTable.lib[resourceIndex];
          resourceName = stringTable.getString(
            resourceTable.name[resourceIndex]
          );
          host =
            hostStringIndex === undefined
              ? null
              : stringTable.getString(hostStringIndex);
          resourceType = resourceTable.type[resourceIndex];
        }
        const lib =
          libIndex === undefined || libIndex === null || libIndex === -1
            ? undefined
            : libs[libIndex];

        return [
          locationName,
          {
            funcName,
            isJS,
            resourceIndex,
            address,
            fileName,
            lineNumber,
            columnNumber,
            libIndex,
            resourceName,
            host,
            resourceType,
            lib,
          },
        ];
      })
    ).toMatchSnapshot();
  });
});

describe('gecko counters processing', function() {
  function setup() {
    // Create a gecko profile with counters.
    const findMainThread = profile =>
      ensureExists(
        profile.threads.find(thread => thread.name === 'GeckoMain'),
        'There should be a GeckoMain thread in the Gecko profile'
      );

    const parentGeckoProfile = createGeckoProfile();
    const [childGeckoProfile] = parentGeckoProfile.processes;

    const parentPid = findMainThread(parentGeckoProfile).pid;
    const childPid = findMainThread(childGeckoProfile).pid;
    expect(parentPid).toEqual(3333);
    expect(childPid).toEqual(2222);

    const parentCounter = createGeckoCounter(
      findMainThread(parentGeckoProfile)
    );
    const childCounter = createGeckoCounter(findMainThread(childGeckoProfile));
    parentGeckoProfile.counters = [parentCounter];
    childGeckoProfile.counters = [childCounter];
    return {
      parentGeckoProfile,
      parentPid,
      childPid,
      parentCounter,
      childCounter,
    };
  }

  it('can extract the counter information correctly', function() {
    const { parentGeckoProfile, parentPid, childPid } = setup();
    const processedProfile = processProfile(parentGeckoProfile);
    const counters = ensureExists(
      processedProfile.counters,
      'Expected to find counters on the processed profile'
    );
    expect(counters.length).toBe(2);
    expect(counters[0].pid).toBe(parentPid);
    expect(counters[1].pid).toBe(childPid);

    const findMainThreadIndexByPid = (pid: number): number =>
      processedProfile.threads.findIndex(
        thread => thread.name === 'GeckoMain' && thread.pid === pid
      );

    expect(counters[0].mainThreadIndex).toBe(
      findMainThreadIndexByPid(parentPid)
    );
    expect(counters[1].mainThreadIndex).toBe(
      findMainThreadIndexByPid(childPid)
    );
  });

  it('offsets the counter timing for child processes', function() {
    const { parentGeckoProfile, parentCounter, childCounter } = setup();
    const processedProfile = processProfile(parentGeckoProfile);
    const processedCounters = ensureExists(processedProfile.counters);

    const originalTime = [0, 1, 2, 3, 4, 5, 6];
    const offsetTime = originalTime.map(n => n + 1000);

    const extractTime = counter =>
      counter.sample_groups.samples.data.map(tuple => tuple[0]);

    // The original times and parent process are not offset.
    expect(extractTime(parentCounter)).toEqual(originalTime);
    expect(extractTime(childCounter)).toEqual(originalTime);
    expect(processedCounters[0].sampleGroups.samples.time).toEqual(
      originalTime
    );

    // The subprocess times are offset when processed:
    expect(processedCounters[1].sampleGroups.samples.time).toEqual(offsetTime);
  });
});

describe('serializeProfile', function() {
  it('should produce a parsable profile string', async function() {
    const profile = processProfile(createGeckoProfile());
    const serialized = serializeProfile(profile);
    expect(JSON.parse.bind(null, serialized)).not.toThrow();
  });

  it('should produce the same profile in a roundtrip', async function() {
    const profile = processProfile(createGeckoProfile());
    const serialized = serializeProfile(profile);
    const roundtrip = await unserializeProfileOfArbitraryFormat(serialized);
    // FIXME: Uncomment this line after resolving `undefined` serialization issue
    // See: https://github.com/firefox-devtools/profiler/issues/1599
    // expect(profile).toEqual(roundtrip);

    const secondSerialized = serializeProfile(roundtrip);
    const secondRountrip = await unserializeProfileOfArbitraryFormat(
      secondSerialized
    );
    expect(roundtrip).toEqual(secondRountrip);
  });
});

describe('sanitizePII', function() {
  function getRemoveProfileInformation(
    customFields: Object
  ): RemoveProfileInformation {
    return {
      shouldRemoveThreads: new Set(),
      shouldRemoveThreadsWithScreenshots: new Set(),
      shouldRemoveNetworkUrls: false,
      shouldRemoveAllUrls: false,
      shouldFilterToCommittedRange: null,
      shouldRemoveExtensions: false,
      ...customFields,
    };
  }
  it('should sanitize the threads if they are provided', function() {
    const profile = processProfile(createGeckoProfile());
    // There are 3 threads in the beginning.
    expect(profile.threads.length).toEqual(3);
    const PIIToRemove = getRemoveProfileInformation({
      shouldRemoveThreads: new Set([0, 2]),
    });

    const sanitizedProfile = sanitizePII(profile, PIIToRemove);
    // First and last threads are removed and now there are only 1 thread.
    expect(sanitizedProfile.threads.length).toEqual(1);
  });

  it('should sanitize the screenshots if they are provided', function() {
    const profile = processProfile(createGeckoProfile());
    // Checking if we have screenshot markers just in case.
    let screenshotMarkerFound = false;
    for (const thread of profile.threads) {
      for (const data of thread.markers.data) {
        if (data && data.type === 'CompositorScreenshot') {
          screenshotMarkerFound = true;
          break;
        }
      }
    }
    expect(screenshotMarkerFound).toEqual(true);

    const PIIToRemove = getRemoveProfileInformation({
      shouldRemoveThreadsWithScreenshots: new Set([0, 1, 2]),
    });

    const sanitizedProfile = sanitizePII(profile, PIIToRemove);
    screenshotMarkerFound = false;
    for (const thread of sanitizedProfile.threads) {
      for (const data of thread.markers.data) {
        if (data && data.type === 'CompositorScreenshot') {
          screenshotMarkerFound = true;
          break;
        }
      }
    }
    expect(screenshotMarkerFound).toEqual(false);
  });

  it('should sanitize the pages information', function() {
    const profile = processProfile(createGeckoProfile());

    for (const page of ensureExists(profile.pages)) {
      expect(page.url.includes('http')).toBe(true);
    }

    const PIIToRemove = getRemoveProfileInformation({
      shouldRemoveNetworkUrls: true,
    });

    const sanitizedProfile = sanitizePII(profile, PIIToRemove);
    for (const page of ensureExists(sanitizedProfile.pages)) {
      expect(page.url.includes('http')).toBe(false);
    }
  });

  it('should sanitize the network URLS', function() {
    const profile = processProfile(createGeckoProfile());
    const PIIToRemove = getRemoveProfileInformation({
      shouldRemoveNetworkUrls: true,
    });

    const sanitizedProfile = sanitizePII(profile, PIIToRemove);
    for (const thread of sanitizedProfile.threads) {
      const stringArray = thread.stringTable.serializeToArray();
      for (let i = 0; i < thread.markers.length; i++) {
        const currentMarker = thread.markers.data[i];
        if (
          currentMarker &&
          currentMarker.type &&
          currentMarker.type === 'Network'
        ) {
          expect(currentMarker.URI).toBeFalsy();
          expect(currentMarker.RedirectURI).toBeFalsy();
          const stringIndex = thread.markers.name[i];
          expect(stringArray[stringIndex].includes('http')).toBe(false);
        }
      }
    }
  });

  it('should sanitize all the URLS', function() {
    const profile = processProfile(createGeckoProfile());
    const PIIToRemove = getRemoveProfileInformation({
      shouldRemoveAllUrls: true,
    });

    const sanitizedProfile = sanitizePII(profile, PIIToRemove);
    for (const thread of sanitizedProfile.threads) {
      const stringArray = thread.stringTable.serializeToArray();
      for (const string of stringArray) {
        // We are keeping the http(s) and removing the rest.
        // That's why we can't test it with `includes('http')`.
        // Tested `.com` here since all of the test urls have .com in it
        expect(string.includes('.com')).toBe(false);
      }
    }
  });

  it('should sanitize extensions', function() {
    const profile = processProfile(createGeckoProfile());
    expect(profile.meta.extensions).not.toEqual(undefined);
    // For flow
    if (profile.meta.extensions !== undefined) {
      const extensions = profile.meta.extensions;
      expect(extensions.length).toEqual(3);
      expect(extensions.id.length).toEqual(3);
      expect(extensions.name.length).toEqual(3);
      expect(extensions.baseURL.length).toEqual(3);
    }
    const PIIToRemove = getRemoveProfileInformation({
      shouldRemoveExtensions: true,
    });

    const sanitizedProfile = sanitizePII(profile, PIIToRemove);
    expect(sanitizedProfile.meta.extensions).not.toEqual(undefined);
    // For flow
    if (sanitizedProfile.meta.extensions !== undefined) {
      const extensions = sanitizedProfile.meta.extensions;
      expect(extensions.length).toEqual(0);
      expect(extensions.id.length).toEqual(0);
      expect(extensions.name.length).toEqual(0);
      expect(extensions.baseURL.length).toEqual(0);
    }
  });
});
