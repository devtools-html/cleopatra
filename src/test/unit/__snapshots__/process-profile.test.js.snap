// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extract functions and resource from location strings extracts the information for all different types of locations 1`] = `
Array [
  Array [
    "0xc0ff33",
    Object {
      "address": 65331,
      "fileName": null,
      "funcName": "0xc0ff33",
      "host": null,
      "isJS": false,
      "lib": Object {
        "arch": "",
        "breakpadId": "",
        "debugName": "",
        "debugPath": "",
        "end": 13631488,
        "name": "No symbols library",
        "offset": 0,
        "path": "",
        "start": 12582912,
      },
      "libIndex": 0,
      "lineNumber": null,
      "resourceIndex": 0,
      "resourceName": "",
      "resourceType": 1,
    },
  ],
  Array [
    "cppFunction1 (in c++ resource name1) + 123",
    Object {
      "address": -1,
      "fileName": null,
      "funcName": "cppFunction1",
      "host": null,
      "isJS": false,
      "lib": undefined,
      "libIndex": -1,
      "lineNumber": null,
      "resourceIndex": 1,
      "resourceName": "c++ resource name1",
      "resourceType": 1,
    },
  ],
  Array [
    "cppFunction2 (in c++ resource name2) (234:345)",
    Object {
      "address": -1,
      "fileName": null,
      "funcName": "cppFunction2",
      "host": null,
      "isJS": false,
      "lib": undefined,
      "libIndex": -1,
      "lineNumber": null,
      "resourceIndex": 2,
      "resourceName": "c++ resource name2",
      "resourceType": 1,
    },
  ],
  Array [
    "cppFunction3 (in c++ resource name2)",
    Object {
      "address": -1,
      "fileName": null,
      "funcName": "cppFunction3",
      "host": null,
      "isJS": false,
      "lib": undefined,
      "libIndex": -1,
      "lineNumber": null,
      "resourceIndex": 2,
      "resourceName": "c++ resource name2",
      "resourceType": 1,
    },
  ],
  Array [
    "jsFunction1 (http://script.com/one.js:456)",
    Object {
      "address": -1,
      "fileName": "http://script.com/one.js",
      "funcName": "jsFunction1",
      "host": "script.com",
      "isJS": true,
      "lib": undefined,
      "libIndex": undefined,
      "lineNumber": 456,
      "resourceIndex": 3,
      "resourceName": "http://script.com",
      "resourceType": 3,
    },
  ],
  Array [
    "http://script.com/one.js:456",
    Object {
      "address": -1,
      "fileName": "http://script.com/one.js",
      "funcName": "(root scope) http://script.com/one.js",
      "host": "script.com",
      "isJS": true,
      "lib": undefined,
      "libIndex": undefined,
      "lineNumber": 456,
      "resourceIndex": 3,
      "resourceName": "http://script.com",
      "resourceType": 3,
    },
  ],
  Array [
    "mysterious location",
    Object {
      "address": -1,
      "fileName": null,
      "funcName": "mysterious location",
      "host": null,
      "isJS": false,
      "lib": undefined,
      "libIndex": undefined,
      "lineNumber": null,
      "resourceIndex": -1,
      "resourceName": null,
      "resourceType": null,
    },
  ],
]
`;
